Breaks associated_type_bounds

Input code:

```rust
#![feature(associated_type_bounds)]

fn f<I: Iterator<Item: Clone>>() {}
```

As of current rustfmt master (e0e2f0db537904b0efa78a106384e62852b806f0), rustfmt applies the following diff which makes the original code no longer compile:

```diff
- fn f<I: Iterator<Item: Clone>>() {}
+ fn f<I: Iterator<Item = Clone>>() {}
```

```console
warning: trait objects without an explicit `dyn` are deprecated
 --> src/main.rs:3:25
  |
3 | fn f<I: Iterator<Item = Clone>>() {}
  |                         ^^^^^ help: use `dyn`: `dyn Clone`
  |
  = note: #[warn(bare_trait_objects)] on by default

error[E0277]: the size for values of type `(dyn std::clone::Clone + 'static)` cannot be known at compilation time
 --> src/main.rs:3:1
  |
3 | fn f<I: Iterator<Item = Clone>>() {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `std::marker::Sized` is not implemented for `(dyn std::clone::Clone + 'static)`
  = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
  = note: required by `std::iter::Iterator`

error[E0038]: the trait `std::clone::Clone` cannot be made into an object
 --> src/main.rs:3:1
  |
3 | fn f<I: Iterator<Item = Clone>>() {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::clone::Clone` cannot be made into an object
  |
  = note: the trait cannot require that `Self : Sized`
```

--- a/src/types.rs
+++ b/src/types.rs
@@ -168,14 +168,22 @@ impl<'a> Rewrite for SegmentParam<'a> {
             SegmentParam::LifeTime(lt) => lt.rewrite(context, shape),
             SegmentParam::Type(ty) => ty.rewrite(context, shape),
             SegmentParam::Binding(assoc_ty_constraint) => {
-                let mut result = match context.config.type_punctuation_density() {
-                    TypeDensity::Wide => {
-                        format!("{} = ", rewrite_ident(context, assoc_ty_constraint.ident))
+                let mut result = match assoc_ty_constraint.kind {
+                    ast::AssocTyConstraintKind::Bound { .. } => {
+                        format!("{}: ", rewrite_ident(context, assoc_ty_constraint.ident))
                     }
-                    TypeDensity::Compressed => {
-                        format!("{}=", rewrite_ident(context, assoc_ty_constraint.ident))
+                    ast::AssocTyConstraintKind::Equality { .. } => {
+                        match context.config.type_punctuation_density() {
+                            TypeDensity::Wide => {
+                                format!("{} = ", rewrite_ident(context, assoc_ty_constraint.ident))
+                            }
+                            TypeDensity::Compressed => {
+                                format!("{}=", rewrite_ident(context, assoc_ty_constraint.ident))
+                            }
+                        }
                     }
                 };
+
                 let budget = shape.width.checked_sub(result.len())?;
                 let rewrite = assoc_ty_constraint
                     .kind

--- a/tests/source/associated_type_bounds.rs
+++ b/tests/source/associated_type_bounds.rs
@@ -1,0 +1,13 @@
+// See #3657 - https://github.com/rust-lang/rustfmt/issues/3657
+
+#![feature(associated_type_bounds)]
+
+fn f<I: Iterator<Item: Clone>>() {}
+
+fn g<I: Iterator<Item : Clone>>() {}
+
+fn h<I: Iterator<Item :      Clone>>() {}
+
+fn i<I: Iterator<Item:Clone>>() {}
+
+fn j<I: Iterator<Item :  Clone+'a>>() {}

--- a/tests/target/associated_type_bounds.rs
+++ b/tests/target/associated_type_bounds.rs
@@ -1,0 +1,13 @@
+// See #3657 - https://github.com/rust-lang/rustfmt/issues/3657
+
+#![feature(associated_type_bounds)]
+
+fn f<I: Iterator<Item: Clone>>() {}
+
+fn g<I: Iterator<Item: Clone>>() {}
+
+fn h<I: Iterator<Item: Clone>>() {}
+
+fn i<I: Iterator<Item: Clone>>() {}
+
+fn j<I: Iterator<Item: Clone + 'a>>() {}

